skxray/core/speckle.py

```diff
@@ -0,0 +1,295 @@
+       spectroscopy with partially coherent x-rays" J. Synchrotron Rad.,
+       vol 21, p 1288-1295, 2014.
+
+    .. [2] R. Bandyopadhyay, A. S. Gittings, S. S. Suh, P.K. Dixon and
+       D.J. Durian "Speckle-visibilty Spectroscopy: A tool to study
+       time-varying dynamics" Rev. Sci. Instrum. vol 76, p  093110, 2005.
+
+    There is an example in https://github.com/scikit-xray/scikit-xray-examples
+    It will demonstrate the use of these functions in this module for
+    experimental data.
+
+    """
+    if max_cts is None:
+        max_cts = roi.roi_max_counts(image_sets, label_array)
+
+    # find the label's and pixel indices for ROI's
+    labels, indices = corr.extract_label_indices(label_array)
+
+    # number of ROI's
+    u_labels = list(np.unique(labels))
+    num_roi = len(u_labels)
+
+    # create integration times
+    time_bin = geometric_series(timebin_num, number_of_img)
+
+    # number of items in the time bin
+    num_times = len(time_bin)
+
+    # number of pixels per ROI
+    num_pixels = np.bincount(labels, minlength=(num_roi+1))[1:]
+
+    # probability density of detecting photons
+    prob_k_all = np.zeros([num_times, num_roi], dtype=np.object)
+
+    # square of probability density of detecting photons
+    prob_k_pow_all = np.zeros_like(prob_k_all)
+
+    # standard deviation of probability density of detecting photons
+    prob_k_std_dev = np.zeros_like(prob_k_all)
+
+    # get the bin edges for each time bin for each ROI
+    bin_edges = np.zeros(prob_k_all.shape[0], dtype=prob_k_all.dtype)
+    for i in range(num_times):
+        bin_edges[i] = np.arange(max_cts*2**i)
+
+    start_time = time.time()  # used to log the computation time (optionally)
+
+    for i, images in enumerate(image_sets):
+        # Ring buffer, a buffer with periodic boundary conditions.
+        # Images must be keep for up to maximum delay in buf.
+        buf = np.zeros([num_times, timebin_num],
+                       dtype=np.object)  # matrix of buffers
+
+        # to track processing each time level
+        track_level = np.zeros(num_times)
+
+        # to increment buffer
+        cur = np.full(num_times, timebin_num)
+
+        # to track how many images processed in each level
+        img_per_level = np.zeros(num_times, dtype=np.int64)
+
+        prob_k = np.zeros_like(prob_k_all)
+        prob_k_pow = np.zeros_like(prob_k_all)
+
+        for n, img in enumerate(images):
+            cur[0] = (1 + cur[0]) % timebin_num
+            # read each frame
+            # Put the image into the ring buffer.
+            buf[0, cur[0] - 1] = (np.ravel(img))[indices]
+
+            _process(num_roi, 0, cur[0] - 1, buf, img_per_level, labels,
+                     max_cts, bin_edges[0], prob_k, prob_k_pow)
+
+            # check whether the number of levels is one, otherwise
+            # continue processing the next level
+            processing = num_times > 1
+            level = 1
+
+            while processing:
```

Comment:
change this to `while level < num_times` and get rid of the `processing` local variable